[c10d] a new Pytorch API (split_group) to create a process group

shuqiangzhang opened the pull request:
This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130507

ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10d] a new Pytorch API (split_group) to create a process group through
ncclCommSplit
Summary:
n current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg, when device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

[ghstack-poisoned]

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    Create a new distributed group splitted from the given parent group.
+
+    users of this API must gurantee that all ranks in the parent group enter this API call.
+    And the split of the group is the same accross the ranks.
```
call, and

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+            the default process group will be used. Users need to gurantee that 
+            the parent group is fully initialized (e.g, communicators are initialized)
+        split_ranks (list[int]): the split ranks, which is a list of list of ranks.
+            Users need to make sure the validity of the split rannks such that one 
```
raanks

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+        group_desc (str, optional): a string to describe the process group.
+
+    Returns:
+        A handle of distributed group that can be given to collective calls or
```
s/handle/class`torch.distributed.ProcessGroup`

(or whatever the right way to refer to a class in docstrings is)

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
 
 
+@_time_logger
+def split_group(
```
please add mypy typing hints

shuqiangzhang replied:
myspy tying hints does not work well with 'None' default value, e.g.,
default has type "None", argument has type "timedelta"

wconstab replied:
it takes some work to get right but i think its usually worthwhile.  In general you can use Optional[] or Union[] to make the None case or the multi-type cases work out.  Then it'll force you to assert its not none or assert its the right type before using it.

wconstab replied:
just double checked, other APIs in the file are using mypy.  if init_process_group can do it then you can too :) 

```
@_time_logger
def init_process_group(
    backend: Optional[str] = None,
    init_method: Optional[str] = None,
    timeout: Optional[timedelta] = None,
    world_size: int = -1,
    rank: int = -1,
    store: Optional[Store] = None,
    group_name: str = "",
    pg_options: Optional[Any] = None,
    device_id: Optional[torch.device] = None,
) -> None:
```

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+        parent_pg (ProcessGroup, optional): The parent process group. If None,
+            the default process group will be used. Users need to gurantee that 
+            the parent group is fully initialized (e.g, communicators are initialized)
+        split_ranks (list[int]): the split ranks, which is a list of list of ranks.
```
typing looks wrong? you say list[int] and then later you say list of list of int.  

also, i'm reading this and not understanding how split_ranks actually works.  need to say what each int represents and if there are nested lists, what each list represents.

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,175 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    pg_options=None,
+    group_desc=None,
+):
+    """
```
should document that some backends do not support this API and what the behavior is in unsupported case

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group through"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+
+    pg._register_backend(torch.device("cuda"), backend_type, backend_class)
+
+    # set group_name and group_dsec to backend
```
typo desc

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    global _world
+    default_pg = _get_default_group()
+    device_id = default_pg.bound_device_id
+    if not device_id:
```
we might need a helper fn `_can_split` since i think we check this logic in multiple places

shuqiangzhang replied:
Will create a helper func, but I think here just checking device_id is more efficient, as we need to assert use device_id in many other cases, such as get_backend and perform_nocolor_split

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+        raise ValueError(f"Group {parent_pg} is not registered")
+
+    parent_global_to_group_ranks = _world.pg_group_ranks[parent_pg]
+    parent_group_to_global_ranks = {
```
dont we have a helper already, like get_group_rank()?

shuqiangzhang replied:
We need convert a subgroup of group ranks to a group of global ranks, so we need a group rank to global rank dict, instead of just group ranks. With parent_global_to_group_ranks  as a start, we just need to reverse the mapping, there is no need to call the helper function

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    if (
+        not parent_backend
+        or not parent_backend.supports_splitting
+        or not isinstance(parent_backend, ProcessGroupNCCL)
```
shouldn't the .supports_splitting check be sufficient? we'll get PRs eventually for widening this from just nccl so if we already have an api maybe just use that?

shuqiangzhang replied:
I think that's fine, this API is only for NCCL. If we need to widen this API, it needs to be overhauled anyway

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+
+    prefix_store = PrefixStore(f"{group_name}/", default_store)
+    base_pg_options = ProcessGroup.Options(backend=str(backend))
+    base_pg_options._timeout = timeout
```
are there any other options we need to copy actually?

shuqiangzhang replied:
This options is just a placeholder to create the base pg I think, the real option is pg_options

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    pg_options.split_color = _process_group_color(my_group)
+    pg_options.global_ranks_in_group = global_ranks_in_my_group
+    pg_options.group_name = group_name
+    backend_class = ProcessGroupNCCL(
```
shouldn't this be dynamic based on the backend_class we extracted above?

shuqiangzhang replied:
I would prefer the code to be short and succinct and this API is only for nccl (not gloo/MPI) and we check the input rigorously. That's why we can keep the code here much more succinct than the nested calls of new_group

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    assert group_name is not None
+    assert group_desc is not None
+    pg._set_group_name(group_name)
+    pg._set_group_desc(group_desc)
```
wonder if we should make a default group_desc if None is provided.  something like parent_desc + "split" + ranks?  that might be too verbose but.. idk

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4228,6 +4251,140 @@ def test_nccl_errors_dump(self):
+        else:
+            self.assertEqual(ng2, c10d.GroupMember.NON_GROUP_MEMBER)
+        # give enough time for the broadcast to finish before destroying all pgs.
+        time.sleep(2)
```
can't you just wait the broadcast and synchronize cuda? then avoid sleep?

shuqiangzhang replied:
weirdly, broadcast() should be sync call by default, which means work.wait is called. But obviously it does not work as expected, this is something we need to follow up

wconstab replied:
No, it works as expected but you forgot the cuda synchronize().

Work.wait() syncs one cuda stream used for comms back to the default cuda stream, it does Not sync to the CPU.

shuqiangzhang replied:
@wconstab do you mean an additional call of torch.cuda.synchronize(device=device)? This did not work in my  and the tests still fail. 

Also in normal pg collectives, Python users are expected to call both wait() and cuda synchronize() if he/she want the collective to be a sync call? This seems bad design for wait()

wconstab replied:
Yes, that's the way cuda/pytorch works to be honest.

Even if you call torch.matmul() and then exit your test, the matmul might still be running on you. You have to explicitly sync gpu if you want to make sure it's finished.

The reason we don't do a sync inside work.wait is because most often that is not what users want and would be slower. Instead it helps ensure that collective finishes before any new you kernels start. It helps manage dependencies between different streams of cuda ops.

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4228,6 +4251,140 @@ def test_nccl_errors_dump(self):
+
+    @requires_nccl_version((2, 18), "Need NCCL 2.18+ for ncclCommSplit")
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @skip_if_lt_x_gpu(8)
```
could you add tests that make sure options from parent group are applied to the split group (timeout, perhaps also high-prio stream, any others that i forgot about?

shuqiangzhang replied:
will add a few assertEqual in the basic test_comm_split_group

fduwjj commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4297,6 +4298,188 @@ def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    _register_process_group(group_name, pg)
+    _world.pg_backend_config[pg] = str(backend_config)
+    pg_tag = f"ptd:{group_name}"
+    _world.tags_to_pg.setdefault("", []).append(pg)
+    _world.tags_to_pg.setdefault(pg_tag, []).append(pg)
```
Why do we need to do this twice?

shuqiangzhang replied:
This is basically copying the same logic of _new_process_group_helper when no tag is provided by users. But I guess, we just need to keep the later

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
> in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API

IIRC, there is a new_subgroup API that requires one instead of 32 calls. Would that help?

shuqiangzhang commented:
Not enough, as the new_subgroup is just a wrapper of new_group and internally it still calls new_group many times, one for each sublist, it can also only create new groups only from default pg as parent group. and it has constraints like, examples  of "Create subgroups of equal size". new_subgroups_by_enumeration is closer to split_group, but still subject to most of above drawbacks

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] a new Pytorch API (split_group) to create a process group"

This is the implementation following the RFC: https://github.com/pytorch/pytorch/issues/130407


ncclCommSplit
Summary:
In current Pytorch/c10d, the new_group API is used to create a new
process group from the default pg.  When device_id is specified in
init_process_group and nccl is used as the backend, the new_group call
will use ncclCommSplit to create the nccl communicators to save
communicator resources. It has a few drawbacks:

Redundant calls
Suppose the default group has 256 ranks, we need to have 32 children PGs
and each child PG has 8 ranks. in this case, each rank needs to call
new_group and ncclCommSplit 32 times because of how we implement
new_group API and the collective requirement of ncclCommSplit. For a
specific global rank, 31 calls of ncclCommSplit would be no_color split,
and only 1 of them is colored split. With the proposed new split_group
API, we expect only 1 call of split_group/ncclCommSplit is needed per
rank in the above example case

new_group can only split from default_pg
Ideally, a new pg should be able to be split from any pg

With the new split_group API, users can create new PGs using
ncclCommSplit with less number of calls and initialize the PG eagerly.
This is also useful in the cases of creating many P2P communicators.
Test Plan:
New UTs:
e.g., python test/distributed/test_c10d_nccl.py -k
test_comm_split_group_larger_scale
Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
